Restore no-crash behaviour when read-only config file(s)

karthanistyr opened the pull request:
Prior to migrating handling configuration with Configurate,
paper did not crash when it found it couldn't overwrite the
existing paper.yml configuration file (but left a stack trace).

After the migration, paper crashes when it is unable to overwrite
any of paper-global.yml, paper-world-defaults.yml or a
world-specific config file.

This aims at returning to the previous behaviour of not
crashing when paper is unable to overwrite a configuration
file.

Signed-off-by: Antoine Mazeas <antoine@karthanis.net>

Fix #8301

karthanistyr committed:
Restore no-crash behaviour when read-only config file(s)

Prior to migrating handling configuration with Configurate,
paper did not crash when it found it couldn't overwrite the
existing paper.yml configuration file (but left a stack trace).

After the migration, paper crashes when it is unable to overwrite
any of paper-global.yml, paper-world-defaults.yml or a
world-specific config file.

This aims at returning to the previous behaviour of not
crashing when paper is unable to overwrite a configuration
file.

Signed-off-by: Antoine Mazeas <antoine@karthanis.net>

Machine-Maker commented on patches/server/0004-Paper-config-files.patch:
```diff
@@ -4265,8 +4278,20 @@ index 7d965247833c91dc824e6cc56e8b0fe5f3413d1d..08ae7a96e93c0d8547f560b3f7538045
+         try {
+             // CraftBukkit start - Don't attempt writing to file if it's read only
+             if (path.toFile().exists() && !path.toFile().canWrite()) {
++                Settings.LOGGER.warn("Can not write to file {}, skipping.", path);
```
Paper comment

karthanistyr replied:
Done

Lulu13022002 replied:
The comment must be in the same line when it's only one line of diff

karthanistyr replied:
and done

Machine-Maker commented on patches/server/0004-Paper-config-files.patch:
```diff
@@ -148,6 +148,7 @@ index 0000000000000000000000000000000000000000..31325994ab441c59a4c0bd9f3f9db3d9
 +
 +public abstract class Configurations<G, W> {
 +
++    private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger();
```
You can use imports here, as this patch is the patch that creates the file. So go ahead and import LogUtils and Logger and use those directly.

We like to avoid adding imports to files that we are changing as that can create useless conflicts on rebases. But if the file is being first created, you can add as many imports as you want as its never going to conflict on a rebase.

karthanistyr replied:
Done

Machine-Maker commented on patches/server/0004-Paper-config-files.patch:
```diff
@@ -217,6 +218,17 @@ index 0000000000000000000000000000000000000000..31325994ab441c59a4c0bd9f3f9db3d9
++        try {
++            loader.save(node);
++        } catch (ConfigurateException ex) {
++            if (ex.getCause() instanceof java.nio.file.AccessDeniedException) {
```
add import

karthanistyr replied:
Done

Lulu13022002 commented on patches/server/0004-Paper-config-files.patch:
```diff
@@ -217,6 +218,17 @@ index 0000000000000000000000000000000000000000..31325994ab441c59a4c0bd9f3f9db3d9
++        } catch (ConfigurateException ex) {
++            if (ex.getCause() instanceof java.nio.file.AccessDeniedException) {
++                LOGGER.warn("Could not save {}: Paper could not persist the full set of configuration settings in the configuration file. Any setting missing from the configuration file will be set with its default value in memory. Admins should make sure to review the configuration documentation at https://docs.papermc.io/paper/configuration for more details.", filename);
++                LOGGER.warn("Exception: {} ", ex);
```
The {} is not interpreted (see your log) because of the type. A lot of logger supports an optional parameter at the end of their function to print directly the stacktrace. In short you can just remove the " {}" at the end and it will still work
But i think you can remove this line totally because we already know that it's an exception no need to write Exception: ...

karthanistyr replied:
yes, this was to benefit from `Logger.warn(String, Throwable)` but apparently this can be compressed into a single call, ref: https://www.slf4j.org/faq.html#paramException

Wouldn't we want to keep logging the stack trace to be consistent with other skips like with spigot.yml ?

Lulu13022002 replied:
Yes you can compress that into a single call, also i agree with the consistency between the other files. Basically when this is a real permission issue (and not a wanted behavior) the stacktrace help a lot

karthanistyr replied:
Done as mentioned

karthanistyr committed:
Fix review comments

Compress logging of exeption stack trace into single call
Add imports where relevant
Add Paper comment where needed

Signed-off-by: Antoine Mazeas <antoine@karthanis.net>

karthanistyr committed:
Fix comment format

Signed-off-by: Antoine Mazeas <antoine@karthanis.net>
